File: packages/shared/src/match.ts
import type { Modality, Tactic, Weather } from "./enums";

export interface TeamSpec {
  name: string;
  serve: number;
  attack: number;
  block: number;
  defense: number;
  reception: number;
  setting: number;
  players: string[];
  chemistry: number;
  morale: number;
  fatigue: number;
  tactic: Tactic;
}

export interface ExhibitionRequest {
  modality: Modality;
  weather?: Weather | null;
  seed?: number | null;
  home: TeamSpec;
  away: TeamSpec;
}

export interface SetScore {
  set_no: number;
  home: number;
  away: number;
}

export interface MatchEvent {
  set_no: number;
  rally_no: number;
  event_type: string;
  side: "home" | "away" | "info";
  text: string;
  athlete?: string | null;
}

export interface MatchResult {
  home_sets: number;
  away_sets: number;
  winner: "home" | "away";
  sets: SetScore[];
  events: MatchEvent[];
}

// Mensagens do WebSocket de narração ao vivo.
export type WsMessage =
  | { type: "start"; home: string; away: string }
  | ({ type: "event" } & MatchEvent)
  | {
      type: "end";
      home_sets: number;
      away_sets: number;
      winner: "home" | "away";
      sets: SetScore[];
    }
  | { type: "error"; detail: string };

File: apps/api/app/schemas/match.py
"""Schemas para simulação de partidas (modo exibição/teste da engine)."""

from pydantic import BaseModel, Field

from app.enums import Modality, Tactic, Weather


class TeamSpec(BaseModel):
    """Forças agregadas de uma equipe (0–100) para a engine."""

    name: str
    serve: float = Field(default=60, ge=0, le=100)
    attack: float = Field(default=60, ge=0, le=100)
    block: float = Field(default=60, ge=0, le=100)
    defense: float = Field(default=60, ge=0, le=100)
    reception: float = Field(default=60, ge=0, le=100)
    setting: float = Field(default=60, ge=0, le=100)
    players: list[str] = Field(default_factory=list)
    chemistry: float = Field(default=50, ge=0, le=100)
    morale: float = Field(default=70, ge=0, le=100)
    fatigue: float = Field(default=0, ge=0, le=100)
    tactic: Tactic = Tactic.BALANCED


class ExhibitionRequest(BaseModel):
    modality: Modality
    weather: Weather | None = None
    seed: int | None = None
    home: TeamSpec
    away: TeamSpec


class SetScoreOut(BaseModel):
    set_no: int
    home: int
    away: int


class MatchEventOut(BaseModel):
    set_no: int
    rally_no: int
    event_type: str
    side: str
    text: str
    athlete: str | None = None


class MatchResultOut(BaseModel):
    home_sets: int
    away_sets: int
    winner: str
    sets: list[SetScoreOut]
    events: list[MatchEventOut]
